manager-rs/dauth-service/src/local/queries.rs

```diff
@@ -0,0 +1,955 @@
+        USER_INFO_OPC_FIELD,
+        USER_INFO_SQN_FIELD,
+    ))
+    .execute(pool)
+    .await?;
+    Ok(())
+}
+
+/// Returns the first for a given id, sorted by rank (seqnum).
+pub async fn get_first_vector(
+    transaction: &mut Transaction<'_, Sqlite>,
+    id: &str,
+) -> Result<SqliteRow, DauthError> {
+    Ok(sqlx::query(&format!(
+        "SELECT * 
+        FROM {0}
+        WHERE {1}=$1
+        ORDER BY {2}
+        LIMIT 1;",
+        AV_TABLE_NAME, AV_ID_FIELD, AV_RANK_FIELD,
+    ))
+    .bind(id)
+    .fetch_one(transaction)
+    .await?)
+}
+
+/// Inserts a vector with the given data.
+/// Returns an error if (id, seqnum) is not unique.
+pub async fn insert_vector(
+    transaction: &mut Transaction<'_, Sqlite>,
+    id: &str,
+    seqnum: i64,
+    xres: &[u8],
+    autn: &[u8],
+    rand: &[u8],
+) -> Result<(), DauthError> {
+    sqlx::query(&format!(
+        "INSERT INTO {0}
+        VALUES ($1,$2,$3,$4,$5)",
+        AV_TABLE_NAME
+    ))
+    .bind(id)
+    .bind(seqnum)
+    .bind(xres)
+    .bind(autn)
+    .bind(rand)
+    .execute(transaction)
+    .await?;
+
+    Ok(())
+}
+
+/// Removes the vector with the (id, seqnum) pair.
+pub async fn remove_vector(
+    transaction: &mut Transaction<'_, Sqlite>,
+    id: &str,
+    seqnum: i64,
+) -> Result<(), DauthError> {
+    sqlx::query(&format!(
+        "DELETE FROM {0}
+        WHERE ({1},{2})=($1,$2)",
+        AV_TABLE_NAME, AV_ID_FIELD, AV_RANK_FIELD,
+    ))
+    .bind(id)
+    .bind(seqnum)
+    .execute(transaction)
+    .await?;
+    Ok(())
+}
+
+/// Inserts a kseaf with a given uuid and value.
+pub async fn insert_kseaf(
+    transaction: &mut Transaction<'_, Sqlite>,
+    uuid: &[u8],
+    value: &[u8],
+) -> Result<(), DauthError> {
+    sqlx::query(&format!(
+        "INSERT INTO {0}
+        VALUES ($1,$2)",
+        KSEAF_TABLE_NAME,
+    ))
+    .bind(uuid)
+    .bind(value)
+    .execute(transaction)
+    .await?;
+
+    Ok(())
+}
+
+/// Deletes a kseaf vaule if found.
+pub async fn delete_kseaf(
+    transaction: &mut Transaction<'_, Sqlite>,
+    uuid: &[u8],
+) -> Result<(), DauthError> {
+    sqlx::query(&format!(
+        "DELETE FROM {0}
+        WHERE {1}=$1",
+        KSEAF_TABLE_NAME, KSEAF_ID_FIELD,
+    ))
+    .bind(uuid)
+    .execute(transaction)
+    .await?;
+
+    Ok(())
+}
+
+/// Returns a kseaf value if found.
+pub async fn get_kseaf(
+    transaction: &mut Transaction<'_, Sqlite>,
+    uuid: &[u8],
+) -> Result<SqliteRow, DauthError> {
+    Ok(sqlx::query(&format!(
+        "SELECT * 
+        FROM {0}
+        WHERE {1}=$1;",
+        KSEAF_TABLE_NAME, KSEAF_ID_FIELD,
+    ))
+    .bind(uuid)
+    .fetch_one(transaction)
+    .await?)
+}
+
+/// Insert user info and replace if exists.
+pub async fn user_info_add(
+    transaction: &mut Transaction<'_, Sqlite>,
+    user_id: &Id,
+    k: &[u8],
+    opc: &[u8],
+    sqn_max: &[u8],
+) -> Result<(), DauthError> {
+    sqlx::query(&format!(
+        "REPLACE INTO {0}
```

Comment:
Your function name calls this "add", but it has replace semantics, which is tricky. I've seen the word "upsert" used for this? You document this in the function docstring so it's probably okay, although might still be better with a different name.

Reply:
I like upsert, I wasn't sure what to name it but I'm a big fan of being descriptive in function names. Will change this.